File: apps/api/src/auth/rbac.ts
import { ForbiddenException } from "@nestjs/common";
import { AuthContext } from "./auth.service";

// RBAC + cách ly theo nước (bài toán khó #1) tại tầng service: một staff chỉ thao tác được
// dữ liệu của ĐÚNG nước mình có vai. GLOBAL_ADMIN là vai duy nhất vượt biên giới.
export function isStaffForCountry(auth: AuthContext, countryId: string, allowed: string[]): boolean {
  return auth.roles.some(
    (r) =>
      (r.role === "GLOBAL_ADMIN" && r.countryId === null) ||
      (allowed.includes(r.role) && r.countryId === countryId),
  );
}

export function assertStaffForCountry(auth: AuthContext, countryId: string, allowed: string[]): void {
  if (!isStaffForCountry(auth, countryId, allowed)) {
    throw new ForbiddenException({
      code: "FORBIDDEN",
      message: "You do not have a staff role for this country.",
    });
  }
}

// GLOBAL_ADMIN là vai DUY NHẤT vượt biên giới — dùng cho nhật ký audit toàn cục (mọi nước).
export function isGlobalAdmin(auth: AuthContext): boolean {
  return auth.roles.some((r) => r.role === "GLOBAL_ADMIN" && r.countryId === null);
}

export function assertGlobalAdmin(auth: AuthContext): void {
  if (!isGlobalAdmin(auth)) {
    throw new ForbiddenException({
      code: "FORBIDDEN",
      message: "Only a global admin can access this resource.",
    });
  }
}

File: apps/api-go/internal/auth/rbac.go
package auth

import (
	"net/http"

	"github.com/dangquangmedia/affiliate-global/apps/api-go/internal/apierr"
)

func IsStaffForCountry(auth Context, countryID string, allowed ...string) bool {
	allowedSet := make(map[string]struct{}, len(allowed))
	for _, role := range allowed {
		allowedSet[role] = struct{}{}
	}
	for _, assignment := range auth.Roles {
		if assignment.Role == "GLOBAL_ADMIN" && assignment.CountryID == nil {
			return true
		}
		if assignment.CountryID != nil && *assignment.CountryID == countryID {
			if _, ok := allowedSet[assignment.Role]; ok {
				return true
			}
		}
	}
	return false
}

func AssertStaffForCountry(auth Context, countryID string, allowed ...string) error {
	if IsStaffForCountry(auth, countryID, allowed...) {
		return nil
	}
	return apierr.New(http.StatusForbidden, "FORBIDDEN", "You do not have a staff role for this country.")
}

func IsGlobalAdmin(auth Context) bool {
	for _, assignment := range auth.Roles {
		if assignment.Role == "GLOBAL_ADMIN" && assignment.CountryID == nil {
			return true
		}
	}
	return false
}

func AssertGlobalAdmin(auth Context) error {
	if IsGlobalAdmin(auth) {
		return nil
	}
	return apierr.New(http.StatusForbidden, "FORBIDDEN", "Only a global admin can access this resource.")
}
